orderer/common/multichannel/registrar.go

```diff
@@ -464,31 +478,97 @@ func (r *Registrar) JoinChannel(channelID string, configBlock *cb.Block, isAppCh
 	}
 
 	//TODO save the join-block in the file repo to make this action crash tolerant.
+	//TODO remove join block & ledger if things go bad below
 
-	ledgerResources, err := r.newLedgerResources(configEnv)
+	ledgerRes, err := r.newLedgerResources(configEnv)
 	if err != nil {
-		//TODO remove join block
 		return types.ChannelInfo{}, errors.Wrap(err, "failed creating ledger resources")
 	}
 
-	joinSupport, err := newChainSupportForJoin(configBlock, r, ledgerResources, r.consenters, r.signer, r.blockcutterMetrics, r.bccsp)
+	ordererConfig, _ := ledgerRes.OrdererConfig()
+	consenter, foundConsenter := r.consenters[ordererConfig.ConsensusType()]
+	if !foundConsenter {
+		return types.ChannelInfo{}, errors.Errorf("failed to find a consenter for consensus type: %s", ordererConfig.ConsensusType())
+	}
+
+	clusterConsenter, ok := consenter.(consensus.ClusterConsenter)
+	if !ok {
+		return types.ChannelInfo{}, errors.New("clusterConsenter is not a consensus.ClusterConsenter")
+	}
+	isMember, err := clusterConsenter.IsChannelMember(configBlock)
 	if err != nil {
-		//TODO remove join block, clean ledger resources
-		return types.ChannelInfo{}, errors.Wrap(err, "failed creating chain support for join")
+		return types.ChannelInfo{}, errors.Wrap(err, "failed to determine cluster membership from join-block ")
```

Comment:
Do you think this is the right approach?

The consensus type on a chain could change over time, and it seems like there's no explicit reason not to allow following a channel which the orderer cannot join?  I would have thought that we simply follow the channel, and, whenever a config change commits, look up the consensus plugin, check to see if it is a cluster type, and if so, check to see if we've been added to the membership (and if so, return to begin consenting).

Basically instead of passing a particular clusterConsenter in, why not pass in a map from consensus types to cluster consenter, and let it resolve the particular checker internally?

Reply:
You mean change over time like in `kafka -> raft` or `raft -> something-new` ? (e.g. `a -> b`)

We can easily support these type of transitions if we have the consenter of type `a` return false; and then the consenter of type `b` return the actual membership test. That will cause the orderer to follow the channel for type `a` and then ether follow or join `b`. 

We currently do not support `kafka` and `solo` at all; since they are deprecated. However going forward that won't be a problem.

With what happens here: this error happens when an `etcdraft` consenter fails to process the consensus metadata of a block with `etcdraft` consensus-type. I think we should throw an error here and reject the join-block. I kind of agree that the error message is a bit misleading - I guess a better one would be "bad join-block, error trying to bla bla...". That is, to indicate that the failure is on the user for supplying a bad block, not the system for failing to process it.